### PR TITLE
MAINT: Improve error message when common type not found.

### LINKED ISSUE #18430: ENH: A more helpful error message, when `np.select` types don't match type of default kwarg
The numpy.select function understandably raises an error when the type of the values in the choicelist doesn't match that of the default argument.

The following code:

```python
conds1 = [True, False, False]
conds2 = [False, True, False]
conds3 = [False, False, True]

dates1 = [np.datetime64('2020-01-02'), np.datetime64('2020-02-03'), np.datetime64('2020-03-04')]
dates2 = [d + np.timedelta64(52, 'W') for d in dates1]
dates3 = [d + np.timedelta64(52, 'W') for d in dates2]

np.select([conds1, conds2, conds3], [dates1, dates2, dates3]) # default=0 by default
```

Will raise:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-38-59398fcf15ab> in <module>
----> 1 np.select([conds1, conds2, conds3], [dates1, dates2, dates3])

<__array_function__ internals> in select(*args, **kwargs)

~\Miniconda3\lib\site-packages\numpy\lib\function_base.py in select(condlist, choicelist, default)
    674     # need to get the result type before broadcasting for correct scalar
    675     # behaviour
--> 676     dtype = np.result_type(*choicelist)
    677 
    678     # Convert conditions to arrays and broadcast conditions and choices

<__array_function__ internals> in result_type(*args, **kwargs)

TypeError: invalid type promotion

```

The above code can be fixed by 

```python
np.select([conds1, conds2, conds3], [dates1, dates2, dates3], default=np.datetime64('NaT'))
```

outputs:

```
array(['2020-01-02', '2021-02-01', '2022-03-02'], dtype='datetime64[D]')
```

as expected

The error message in the first example, is quite obscure, especially to someone who is not aware of the default argument in the first place, since it is often omitted. It would be better if the error message was more descriptive, and hinted at the fact that the TypeError is happening due to a mismatch between the type of default, and the type of the values in the choicelists. 

Possible message:

```
TypeError: type mismatch between default parameter and type of choicelists: 'int' != 'numpy.datetime64'
```

### PR DESCRIPTION
**Pull request by terfilip**
Fixes gh-18430

### CODE REVIEW

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -671,11 +671,23 @@ def select(condlist, choicelist, default=0):
         raise ValueError("select with an empty condition list is not possible")
 
     choicelist = [np.asarray(choice) for choice in choicelist]
-    choicelist.append(np.asarray(default))
+
+    try:
+        intermediate_dtype = np.result_type(*choicelist)
+    except TypeError as e:
+        raise TypeError('Choicelist elements do not have a common dtype: {}'
+                        .format(e))
```
Comment (rgommers) [inline numpy/lib/function_base.py:679]: can you use exception chaining here and further down? i.e. `raise TypeError(...) from e`

we get a lot of PRs to add exception chaining (`from e` or `from None`) so we may as well do that whenever we add new code.

**Comment by rgommers** [comment]
@seberg you commented in gh-18430 about a more specific error message - are you happy with this PR?

**Comment by seberg** [comment]
The main advantage will happen in C, that we will print out all dtypes passed to `result_type` (I think my branch already does that, but not quite sure).  But I am OK with pointing to the `choicelist` explicitly, you could even say `choicelist and default value` since the fact that the default has an influcence is the confusing thing.

I think there is a deeper issue here though (basically a bug, that `select` is written only for numeric types).  The `default` is "bad", since the logic would be better the other way around: get common-dtype of all "choicelist arguments", and then create the default based on that (if it doesn't exist).  For `datetime64` that would be: `np.dtype("M8[ns]").type()` which gives `NaT`.

The main issue is that this would actually change behaviour for `object` dtype (and bool, which currently gets upcast).  So it probably would need a `FutureWarning`, etc.  And someone using `object` dtype may have to start passing `default=0` explicitly.

File: numpy/lib/function_base.py (hunk 1 above)
Comment (seberg) [reply numpy/lib/function_base.py:679]: Right, it should have a `from None` if we include the old message in the new error.  Using an `f` string is a bit nicer, i.e. `f"...: {e}"`.

We don't do this type of replacement a lot, but it does make things clearer, so happy to put this in.

**Comment by charris** [comment]
close/reopen

**Comment by charris** [comment]
@terfilip Ping.

**Comment by terfilip** [comment]
I've made a commit with the stylistic changes to the TypeErrors that were suggested.

Now that I've read @seberg 's later comments I can see it makes more sense to infer the default from the dtype of the choicelists. I can work on the change in this PR as well

**Comment by seberg** [comment]
Didn't have new look again, but while I would love to move the default into that direction, there is bound to be some trickier things there and I think we should do it in a different PR.

**Comment by terfilip** [comment]
Okay in that case I have no more changes to make here, unless there are any further suggestions or issues to resolve. I'll explore the  how the default changes could be implemented and make another PR for it later.

File: numpy/lib/function_base.py [hunk 2]
```diff
@@ -671,11 +671,25 @@ def select(condlist, choicelist, default=0):
...
 
     choicelist = [np.asarray(choice) for choice in choicelist]
-    choicelist.append(np.asarray(default))
+
+    try:
+        intermediate_dtype = np.result_type(*choicelist)
+    except TypeError as e:
+        raise TypeError(
+                f'Choicelist elements do not have a common dtype: {e}'
+            ) from None
```
[the remark this answers is not in the document]
Comment (terfilip) [reply numpy/lib/function_base.py]: changed in both places

**Comment by charris** [comment]
Thanks @terfilip .

**Pull request merged**